File: MATTHEW/leetcode/AIcoding/1.Gemini-2.5/javascript_solutions/3372-maximize-the-number-of-target-nodes-after-connecting-trees-i.js
var maximizeTargetNodes = function(n, edges, isTarget, k) {
    const adj = Array(n).fill(0).map(() => []);
    for (const [u, v] of edges) {
        adj[u].push(v);
        adj[v].push(u);
    }

    const visited = Array(n).fill(false);
    const componentTargetCounts = [];

    for (let i = 0; i < n; i++) {
        if (!visited[i]) {
            let currentComponentTargets = 0;
            const stack = [i];
            visited[i] = true;

            while (stack.length > 0) {
                const node = stack.pop();
                if (isTarget[node]) {
                    currentComponentTargets++;
                }

                for (const neighbor of adj[node]) {
                    if (!visited[neighbor]) {
                        visited[neighbor] = true;
                        stack.push(neighbor);
                    }
                }
            }
            componentTargetCounts.push(currentComponentTargets);
        }
    }

    componentTargetCounts.sort((a, b) => b - a);

    let maxTargets = 0;
    const componentsToConnect = Math.min(componentTargetCounts.length, k + 1);
    for (let i = 0; i < componentsToConnect; i++) {
        maxTargets += componentTargetCounts[i];
    }

    return maxTargets;
};

File: MATTHEW/leetcode/AIcoding/1.Gemini-2.5/rust_solutions/3372-maximize-the-number-of-target-nodes-after-connecting-trees-i.rs
struct Solution;

impl Solution {
    pub fn maximize_the_number_of_target_nodes_after_connecting_trees_i(
        n: i32,
        edges: Vec<Vec<i32>>,
        target: Vec<i32>,
    ) -> i32 {
        let n = n as usize;
        let mut adj: Vec<Vec<usize>> = vec![vec![]; n];
        for edge in edges {
            let u = edge[0] as usize;
            let v = edge[1] as usize;
            adj[u].push(v);
            adj[v].push(u);
        }

        let mut visited: Vec<bool> = vec![false; n];
        let mut component_target_counts: Vec<i32> = Vec::new();

        for i in 0..n {
            if !visited[i] {
                let mut current_component_target_count = 0;
                let mut stack: Vec<usize> = Vec::new();
                stack.push(i);
                visited[i] = true;

                while let Some(node) = stack.pop() {
                    if target[node] == 1 {
                        current_component_target_count += 1;
                    }

                    for &neighbor in &adj[node] {
                        if !visited[neighbor] {
                            visited[neighbor] = true;
                            stack.push(neighbor);
                        }
                    }
                }
                component_target_counts.push(current_component_target_count);
            }
        }

        component_target_counts.sort_unstable_by(|a, b| b.cmp(a));

        if component_target_counts.len() <= 1 {
            component_target_counts.get(0).unwrap_or(&0).clone()
        } else {
            component_target_counts[0] + component_target_counts[1]
        }
    }
}